README.md
# ox3apiclient

A small class to help connect to the OpenX Enterprise API. As of version 0.5.0 it  uses
[requests_oauthlib](https://github.com/requests/requests-oauthlib) instead of oauth2.

It currently supports Python 2.6 - 3.5.

As of version 0.4.0, ox3apiclient supports API v2. If your instance is v2,
set the api_path option to "/ox/4.0".

As of version 0.5.0 the client.request method returns a requests.Response object instead of
urllib2.Response and throws a requests.exceptions.HTTPError instead of urllib2.HTTPError.
In addition debugging is now available via the standard python logging facility.

See the [requests documentation](http://docs.python-requests.org/en/latest/) for details.

Basic usage:

````python
import ox3apiclient
import logging

ox = ox3apiclient.client_from_file().logon()

ox.logger.setLevel(logging.DEBUG)
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)
ox.logger.addHandler(ch)

accounts = ox.get('/account')

order = {
    'status': 'Active',
    'name': 'OX3APIClient Object Creation Test',
    'account_uid': accounts['objects'][0]['account_uid'],
    'start_date': '2016-06-01 00:00:00'}

new_order = ox.post('/order', data=order)

ox.delete('/order/%s' % new_order['uid'])

ox.logoff()
````


## Installation

From Pypi

The last released version (from the master branch) is available at [PyPi](http://pypi.python.org/pypi)
````
$ pip install ox3apiclient
````


From Github:

Just clone our git repo:

````
$ git clone https://github.com/openx/OX3-Python-API-Client.git
````

Install the downloaded library:
````
python setup.py install
````
this will install the current dependencies.

## Releasing 
````shell
pip3 install twine wheel
python3 setup.py bdist_wheel
twine upload --repository-url https://us-central1-pypi.pkg.dev/ox-registry-prod/ox-pypi-release --verbose dist/*

````

## Authentication

The recommended method of authentication is to use `ox3apiclient.client_from_file`.
By default this will look for a file named `.ox3rc` in the current current
directory, but this can be overwritten by specifying a `file_path` parameter. The
file should be in the following format:

````
[ox3apiclient]
envs=
    dev
    prod

[dev]
email: you@example.com
password: password123
domain: dev.uidomain.com
realm: dev.uidomain_realm
consumer_key: 1fc5c9ae...
consumer_secret: 7c664d68...
authorization_url: http://custom_sso.uidomain.com/api/index/initiate

[prod]
email: you@example.com
password: password123
domain: uidomain.com
realm: uidomain_realm
consumer_key: 1fc5c9ae...
consumer_secret: 7c664d68...
````

`ox3apiclient.client_from_file` will use the first `env` by default but this can
be overwritten by setting the `env` parameter. If your email and password are set
in `.ox3rc` you can simply chain a call to `logon()`.

Alternatively you can set everything in your code.
````python
email = 'you@example.com'
password = 'password123'
domain = 'uidomain.com'
realm = 'uidomain_realm'
consumer_key = '1fc5c9ae...'
consumer_secret = '7c664d68...'

ox = ox3apiclient.Client(
    email=email,
    password=password,
    domain=domain,
    realm=realm,
    consumer_key=consumer_key,
    consumer_secret=consumer_secret)

ox.logon(email, password)
````

    # To run these tests. Install nose (pip install nose)
    # and run nosetests -sxv tests/  from the root dir

## Tests

Install nose

````bash
pip install nose
```` 
 
 and run the following command line from the root:

````bash
nosetests -sxv tests/
````

History.md

0.6.1 / 2021-09-07
=================
  * Migration to Google Artifact Registry

0.6.1 / 2019-07-23
=================
  * Add ODS endpoint to supported endpoint list

0.6.0 / 2017-06-19
==================
 * Added: Python 3 compatibility
 * Fixed: Non running tests

0.5.2 / 2017-02-02
==================
 * Fixed: packaging for Pypi
 * Added: license information

0.5.0 / 2015-07-26
==================
 * Replaced: urllib2 with requests package
 * Replaced: oauth2 with requests_oauthlib package
 * Added: optional timeout parameter
 * Removed: support for Python 2.4/2.5

0.4.0 / 2013-07-23
==================
 * Added handling for API v2
 * Refined docstrings throughout code

0.3.1 / 2013-06-04
==================
 * Removed: Realm Support
 * Realm can still be given, but has no effect
 * Fixed: Unicode encoding error in REST POSTing

0.3.0 / 2013-03-18
==================
 * Added: HTTP and HTTPS proxy support.
 * Fixed: JSON support for Python 2.4 and 2.5
 * Fixed: OAuth support for Python 2.4
 * Fixed: Version number below (0.2.1)
 

0.2.1 / 2012-09-25
==================
  # Added: `upload_creative` to support uploading creative files.

0.2.0 / 2012-08-29
==================

  * Fixed: JSON parse error when deleting objects with call to `Client.delete()`
  * Added: "Official" support for Python 2.4, 2.5, 2.6
  * Added: `logon` and `logoff` convenience methods.
  * Added: `client_from_file` to load credentials from a config file.
  * `OX3APIClient` is deprecated; use `Client` instead.

0.1.0 / 2012-08-26
==================

  * "Official release"


test.py
import ox3apiclient
import logging

ox = ox3apiclient.client_from_file().logon()


ox.logger.setLevel(logging.DEBUG)
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)
ox.logger.addHandler(ch)

accounts = ox.get('/account')

order = {
    'status': 'Active',
    'name': 'OX3APIClient Object Creation Test',
    'account_uid': accounts['objects'][0]['account_uid'],
    'start_date': '2016-06-01 00:00:00'}

new_order = ox.post('/order', data=order)

ox.delete('/order/%s' % new_order['uid'])

ox.logoff()


test_ox3.py
import ox3apiclient
import logging

ox = ox3apiclient.client_from_file().logon()

ox.logger.setLevel(logging.DEBUG)
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)
ox.logger.addHandler(ch)

accounts = ox.get('/account')

report_get = ox.report_get('/data/1.0/report/fields')

report = {
"startDate": "20190709",
"endDate": "20190716",
"timezone": "UTC",
"attributes" :[
    {
      "id": "day"
    }
    ],
"metrics":[
    {"id": "marketRequests"
    }
    ]
}

report_out = ox.report_post('/data/1.0/report', data=report)


order = {
    'status': 'Active',
    'name': 'OX3APIClient Object Creation Test',
    'account_uid': accounts['objects'][0]['account_uid'],
    'start_date': '2016-06-01 00:00:00'}

new_order = ox.post('/order', data=order)

ox.delete('/order/%s' % new_order['uid'])

ox.logoff()


tests/client.py
# -*- coding: utf-8 -*-
import ox3apiclient
import unittest
from mock import Mock, patch
import os


class TestClient(unittest.TestCase):

    def setUp(self):
        self.email = 'you@example.com'
        self.password = 'password123'
        self.domain = 'uidomain.com'
        self.realm = 'uidomain_realm'
        self.consumer_key = '1fc5c9ae...'
        self.consumer_secret = '7c664d68...'
        self.request_token_url = 'https://ex-sso.openx.org/api/index/initiate'
        self.access_token_url = 'https://ex-sso.openx.org/api/index/token'
        self.authorization_url = 'https://ex-sso.openx.org/api/login/process'
        self.api_path_v1 = '/ox/3.0'
        self.api_path_v2 = '/ox/4.0'
        self.url = 'https://www.example.com'

    def _get_client(self, get_return=None, put_return=None, post_return=None, options_return=None, delete_return=None):
        ex_resp = self._build_mock_response()

        with patch('ox3apiclient.requests.Session') as mock_requests_session:
            with patch('ox3apiclient.Client.log_request') as mock_client_log_request:
                mock_requests_session.return_value.get.return_value = get_return or ex_resp
                mock_requests_session.return_value.post.return_value = post_return or ex_resp
                mock_requests_session.return_value.put.return_value = put_return or ex_resp
                mock_requests_session.return_value.options.return_value = options_return or ex_resp
                mock_requests_session.return_value.delete.return_value = delete_return or ex_resp

                mock_client_log_request.return_value = None
                return ox3apiclient.Client(
                    email=self.email,
                    password=self.password,
                    domain=self.domain,
                    realm=self.realm,
                    consumer_key=self.consumer_key,
                    consumer_secret=self.consumer_secret,
                    request_token_url=self.request_token_url,
                    access_token_url=self.access_token_url,
                    authorization_url=self.authorization_url)

    def _build_mock_response(self):
        resp = Mock()
        resp.request.headers = {'rheader1': 'rvalue1',
                                'rheader2': 'rvalue2'}
        resp.headers = {'header1': 'value1',
                        'header2': 'value2'}
        resp.text = 'oauth_token=key&oauth_token_secret=secret&oauth_callback_confirmed=true'
        resp.json.return_value = {'key1': 'value1',
                                  'key2': 'value2',
                                  'key3': 'value3'}

        resp.status_code = 200
        return resp

    def test_fetch_request_token(self):
        # Authorized Case
        resp = self._build_mock_response()
        client = self._get_client(post_return=resp)

        ret_val = client.fetch_request_token()
        self.assertTrue(isinstance(ret_val, dict))
        self.assertEqual((ret_val['secret'], ret_val['key']), ('secret', 'key'))

        # Unauthorized Case
        resp.status_code = 401
        with self.assertRaises(ox3apiclient.OAuthException):
            client.fetch_request_token()

    def test_authorize_token(self):
        # mock the post response, and do some setup
        r = self._build_mock_response()
        r.text = 'oauth_verifier=verifier'
        r.return_value = {'key': 'key', 'secret': 'secret'}
        client = self._get_client(post_return=r)
        client._token = {'key': 'key', 'secret': 'secret'}

        # Unauthorized Case
        r.status_code = 401
        with self.assertRaises(ox3apiclient.OAuthException):
            client.authorize_token()

        # Authorized Case
        r.status_code = 200
        client.authorize_token()
        self.assertEqual(client._token['verifier'], 'verifier')

    def test_fetch_access_token(self):
        # mock the OAuth1 and session post response
        r = self._build_mock_response()
        r.text = 'oauth_token=key'
        r.return_value = 'oauth_token=key'
        client = self._get_client(post_return=r)
        client._token = {'key': 'key',
                         'secret': 'secret',
                         'verifier': 'verifier'}

        # Unauthorized Case
        r.status_code = 401
        with self.assertRaises(ox3apiclient.OAuthException):
            client.fetch_access_token()

        # Authorized Case
        r.status_code = 200
        self.assertEqual(client.fetch_access_token(), 'key')

    def test_validate_session(self):
        client = self._get_client()

        ret_val = client.validate_session()
        self.assertEqual(ret_val,
                         'oauth_token=key&'
                         'oauth_token_secret=secret&'
                         'oauth_callback_confirmed=true')

    def test_logon(self):
        with patch('ox3apiclient.Client.fetch_request_token') as mock_fetch_request_token:
            with patch('ox3apiclient.Client.authorize_token') as mock_authorize_token:
                with patch('ox3apiclient.Client.fetch_access_token') as mock_fetch_access_token:
                    with patch('ox3apiclient.Client.validate_session') as mock_validate_session:
                        client = self._get_client()
                        mock_fetch_request_token.return_value = None
                        mock_authorize_token.return_value = None
                        mock_fetch_access_token.return_value = None
                        mock_validate_session.return_value = None
                        ret_val = client.logon()
                        self.assertTrue(isinstance(ret_val, ox3apiclient.Client))

    def test_logoff(self):
        client = self._get_client()
        ret_val = client.logoff()
        self.assertTrue(isinstance(ret_val, ox3apiclient.Client))

    def test_get(self):
        client = self._get_client()
        ret_val = client.get(self.url)
        self.assertEqual(ret_val, {'key1': 'value1',
                                   'key2': 'value2',
                                   'key3': 'value3'})

    def test_options(self):
        client = self._get_client()
        ret_val = client.options(self.url)
        self.assertEqual(ret_val, {'key1': 'value1',
                                   'key2': 'value2',
                                   'key3': 'value3'})

    def test_put(self):
        client = self._get_client()
        ret_val = client.put(self.url, data={'k': 'v'})
        self.assertEqual(ret_val, {'key1': 'value1',
                                   'key2': 'value2',
                                   'key3': 'value3'})

    def test_post(self):
        client = self._get_client()
        ret_val = client.post(self.url, data={'k': 'v'})
        self.assertEqual(ret_val, {'key1': 'value1',
                                   'key2': 'value2',
                                   'key3': 'value3'})

    def test_delete(self):
        r = self._build_mock_response()
        r.status_code = 204
        client = self._get_client(delete_return=r)
        ret_val = client.delete('https://example.com')
        self.assertEqual(ret_val, [])

        r.status_code = 200
        r.json.return_value = {'key': 'value'}
        ret_val = client.delete('https://example.com')
        self.assertEqual(ret_val, {'key': 'value'})

    def test_upload_creative(self):
        file_path = os.path.join(os.path.dirname(__file__), 'ox3rctest')
        client = self._get_client()
        ret_val = client.upload_creative('123456789', file_path)
        self.assertEqual(ret_val, {'key1': 'value1',
                                   'key2': 'value2',
                                   'key3': 'value3'})


if __name__ == '__main__':
    # run this using python -m unittest -v tests from the root dir
    unittest.main()


ox3apiclient/__init__.py
# -*- coding: utf-8 -*-

from six.moves import configparser as ConfigParser
from six.moves import http_cookiejar as cookielib
import logging
import mimetypes
from pprint import pformat
import random
import json
from six.moves.urllib.parse import parse_qs, urlparse
import requests
from requests.sessions import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry
from requests_oauthlib import OAuth1

__version__ = '0.6.2'

REQUEST_TOKEN_URL = 'https://sso.openx.com/api/index/initiate'
ACCESS_TOKEN_URL = 'https://sso.openx.com/api/index/token'
AUTHORIZATION_URL = 'https://sso.openx.com/login/process'
API_PATH_V1 = '/ox/3.0'
API_PATH_V2 = '/ox/4.0'
API_PATH_SSO = '/api'
ODS_PATH_V1 = '/data/1.0'
ACCEPTABLE_PATHS = (API_PATH_V1, API_PATH_V2, API_PATH_SSO, ODS_PATH_V1)
JSON_PATHS = (API_PATH_V2,ODS_PATH_V1)
HTTP_METHOD_OVERRIDES = ['DELETE', 'PUT', 'OPTIONS']


class UnknownAPIFormatError(ValueError):
    """Client is passed an unrecognized API path that it cannot handle."""
    pass


class OAuthException(Exception):
    """Client encountered an Oauth error."""
    pass


class Client(object):
    """Client for making requests to the OX3 API. Maintains
    authentication and points all requests at a domain+path
    combination. Handles request and response data in the form
    of Python dictionaries, translated to and from the JSON and
    query string encoding the API itself uses.

    """

    def __init__(self, domain, realm, consumer_key, consumer_secret,
                 callback_url='oob',
                 scheme='http',
                 request_token_url=REQUEST_TOKEN_URL,
                 access_token_url=ACCESS_TOKEN_URL,
                 authorization_url=AUTHORIZATION_URL,
                 api_path=API_PATH_V1,
                 email=None,
                 password=None,
                 http_proxy=None,
                 https_proxy=None,
                 headers=None,
                 timeout=None):
        """

        domain -- Your UI domain. The API is accessed off this domain.
        realm -- This is no longer used. Just specify None.
        consumer_key -- Your consumer key.
        consumer_secret -- Your consumer secret.
        callback_url -- Callback URL to redirect to on successful authorization.
            We default to 'oob' for headless login.
        request_token -- Only override for debugging.
        access_token -- Only override for debugging.
        authorization_url -- Only override for debugging.
        api_path -- Only override for debugging.
        http_proxy -- Optional proxy to send HTTP requests through.
        headers -- list of headers to send with the request
        timeout -- http request timeout in seconds.
        """

        self.domain = domain
        self.consumer_key = consumer_key
        self.consumer_secret = consumer_secret
        self.callback_url = callback_url
        self.scheme = scheme
        self.request_token_url = request_token_url
        self.access_token_url = access_token_url
        self.authorization_url = authorization_url
        self.api_path = api_path
        self.timeout = timeout

        # Validate API path:
        if api_path not in ACCEPTABLE_PATHS:
            msg = '"{}" is not a recognized API path.'.format(api_path)
            msg += '\nLegal paths include:'
            for i in ACCEPTABLE_PATHS:
                msg += '\n{}'.format(i)
            raise UnknownAPIFormatError(msg)

        # These get cleared after log on attempt.
        self._email = email
        self._password = password

        # You shouldn't need to access the token and session objects directly so we'll keep them private.
        self._token = None

        retry_strategy = Retry(connect=5,  # number of retries if we can't connect to server
                               read=5,  # number of retries if we can't read response
                               status_forcelist=(502, ),  # we'll retry when status will be 502 (e.g API deploy)
                               backoff_factor=0.2)  # backoff factor for retry (0.0, 0.5, 1.5)
        adapter = HTTPAdapter(max_retries=retry_strategy)
        # new session with mounting for protocols
        self._session = requests.Session()
        self._session.mount('http://', adapter)
        self._session.mount('https://', adapter)

        # set supplied headers and proxies
        if headers:
            self._session.headers.update(headers)
        if http_proxy:
            self._session.proxies.update({'http': http_proxy})
        if https_proxy:
            self._session.proxies.update({'https': https_proxy})

        self.logger = logging.getLogger(__name__)

    def log_request(self, response):
        self.logger.debug('====={0:=<45}'.format('OX3 api call started'))
        self.logger.debug("%s %s" % (response.request.method, response.request.url))
        self.logger.debug('====={0:=<45}'.format('OX3 api call request headers'))
        for k, v in response.request.headers.items():
            self.logger.debug("%s: %s" % (k, v))
        self.logger.debug('====={0:=<45}'.format('OX3 api call request body'))
        self.logger.debug("%s" % response.request.body)
        self.logger.debug('====={0:=<45}'.format('OX3 api call response headers'))
        for k, v in response.headers.items():
            self.logger.debug("%s: %s" % (k, v))
        self.logger.debug('====={0:=<45}'.format('OX3 api call response body'))
        try:
            self.logger.debug(pformat(json.loads(response.text)))
        except ValueError:
            self.logger.debug("%s" % response.text)
        self.logger.debug('====={0:=<45}'.format('OX3 api call finished'))

    def request(self, url, method='GET', headers=None, data=None, sign=False,
                send_json=False):
        """Helper method to make a (optionally OAuth signed) HTTP request."""

        if headers is None:
            headers = {}

        if sign:
            oauth = OAuth1(client_key=self.consumer_key,
                           resource_owner_key=self._token,
                           callback_uri=self.callback_url,
                           signature_type='query')
        else:
            oauth = None
        if send_json:
            response = self._session.request(method, self._resolve_url(url), headers=headers,
                                             json=data, auth=oauth, timeout=self.timeout)
        else:
            response = self._session.request(method, self._resolve_url(url), headers=headers,
                                             data=data, auth=oauth, timeout=self.timeout)
        self.log_request(response)
        response.raise_for_status()
        return response

    def fetch_request_token(self):
        """Helper method to fetch and set request token.

        Returns token string.
        """
        oauth = OAuth1(client_key=self.consumer_key,
                       client_secret=self.consumer_secret,
                       callback_uri=self.callback_url,
                       signature_type='auth_header')
        response = self._session.post(url=self.request_token_url, auth=oauth, timeout=self.timeout)
        self.log_request(response)
        if response.status_code != 200:
            raise OAuthException("OAuth token request failed (%s) %s" % (response.status_code, response.text))
        credentials = parse_qs(response.text)
        self._token = {'key': credentials['oauth_token'][0],
                       'secret': credentials['oauth_token_secret'][0]}
        return self._token

    def authorize_token(self, email=None, password=None):
        """Helper method to authorize."""
        # Give precedence to credentials passed in methods calls over those set
        # in the instance. This allows you to override user creds that may have
        # been loaded from a file.
        if not email:
            email = self._email

        if not password:
            password = self._password

        if not email or not password:
            raise Exception('Missing email or password')

        data = {
            'email': email,
            'password': password,
            'oauth_token': self._token['key']}

        response = self._session.post(url=self.authorization_url, data=data, timeout=self.timeout)
        self.log_request(response)
        if response.status_code != 200:
            raise OAuthException("OAuth login failed (%s) %s" % (response.status_code, response.text))

        # set token verifier
        self._token['verifier'] = parse_qs(response.text)['oauth_verifier'][0]

    def fetch_access_token(self):
        """Helper method to fetch and set access token.

        Returns token string.
        """
        oauth = OAuth1(client_key=self.consumer_key,
                       client_secret=self.consumer_secret,
                       resource_owner_key=self._token['key'],
                       resource_owner_secret=self._token['secret'],
                       verifier=self._token['verifier'],
                       callback_uri=self.callback_url,
                       signature_type='auth_header')
        response = self._session.post(url=self.access_token_url, auth=oauth, timeout=self.timeout)
        self.log_request(response)
        if response.status_code != 200:
            raise OAuthException("OAuth token verification failed (%s) %s" % (response.status_code, response.text))
        self._token = parse_qs(response.text)['oauth_token'][0]
        return self._token

    def validate_session(self):
        """Validate an API session."""

        # We need to store our access token as the openx3_access_token cookie.
        # This cookie will be passed to all future API requests.
        cookie = cookielib.Cookie(
            version=0,
            name='openx3_access_token',
            value=self._token,
            port=None,
            port_specified=False,
            domain=self.domain,
            domain_specified=True,
            domain_initial_dot=False,
            path='/',
            path_specified=True,
            secure=False,
            expires=None,
            discard=False,
            comment=None,
            comment_url=None,
            rest={})
        self._session.cookies.set_cookie(cookie)

        # v2 doesn't need this extra step, just the cookie:
        if self.api_path == API_PATH_V1:
            response = self._session.put(url=self._resolve_url('/a/session/validate'), timeout=self.timeout)
            self.log_request(response)
            return response.text

    def logon(self, email=None, password=None):
        """Returns self after authentication.

        Single call to complete OAuth login process.

        Keyword arguments:
        email -- user email address.
        password -- user password.

        """
        self.fetch_request_token()
        self.authorize_token(email=email, password=password)
        self.fetch_access_token()
        self.validate_session()
        return self

    def logoff(self):
        """Returns self after deleting authenticated session."""
        if self.api_path == API_PATH_V1:
            response = self._session.delete(self._resolve_url('/a/session'), timeout=self.timeout)
        elif self.api_path == API_PATH_V2:
            response = self._session.delete(self._resolve_url('/session'), timeout=self.timeout)
        elif self.api_path == API_PATH_SSO:
            oauth = OAuth1(client_key=self.consumer_key,
                           resource_owner_key=self._token,
                           callback_uri=self.callback_url,
                           signature_type='query')

            response = self._session.delete(url=self.access_token_url, auth=oauth, timeout=self.timeout)
            if response.status_code != 204:
                raise OAuthException("OAuth token deletion failed (%s) %s" % (response.status_code, response.text))
        else:
            raise UnknownAPIFormatError(
                'Unrecognized API path: %s' % self.api_path)
        self.log_request(response)
        return self

    def _resolve_url(self, url):
        """Converts an API path shorthand into a full URL unless
        given a full url already.

        """
        parse_res = urlparse(url)

        # 2.4 returns a tuple instead of ParseResult. Since ParseResult is a
        # subclass or tuple we can access URL components similarly across
        # 2.4 - 2.7. Yay!

        # If there is no scheme specified we create a fully qualified URL.
        if not parse_res[0]:
            url = '%s://%s%s%s' % (self.scheme, self.domain, self.api_path,
                                   parse_res[2])
            if parse_res[4]:
                url = url + '?' + parse_res[4]

        return url

    def _response_value(self, response):
        """ Utility method. Returns decoded json. If the response content cannot be decoded, then
        the text is returned.

        """
        try:
            return response.json()
        except ValueError:
            return response.text

    def get(self, url, params=None):
        """Issue a GET request to the given URL or API shorthand

        """
        response = self._session.get(self._resolve_url(url), params=params, timeout=self.timeout)
        self.log_request(response)
        response.raise_for_status()
        return self._response_value(response)

    def options(self, url):
        """Send a request with HTTP method OPTIONS to the given
        URL or API shorthand.

        OX3 v2 uses this method for showing help information.

        """
        response = self._session.options(self._resolve_url(url), timeout=self.timeout)
        self.log_request(response)
        response.raise_for_status()
        return self._response_value(response)

    def put(self, url, data=None):
        """Issue a PUT request to url (either a full URL or API
        shorthand) with the data.

        """
        if self.api_path in JSON_PATHS:
            response = self._session.put(self._resolve_url(url), data=json.dumps(data), timeout=self.timeout)
        else:
            response = self._session.put(self._resolve_url(url), data=data, timeout=self.timeout)
        self.log_request(response)
        response.raise_for_status()
        return self._response_value(response)

    def post(self, url, data=None):
        """Issue a POST request to url (either a full URL or API
        shorthand) with the data.

        """
        if self.api_path in JSON_PATHS:
            response = self._session.post(self._resolve_url(url), data=json.dumps(data), timeout=self.timeout)
        else:
            response = self._session.post(self._resolve_url(url), data=data, timeout=self.timeout)
        self.log_request(response)
        response.raise_for_status()
        return self._response_value(response)

    def delete(self, url):
        """Issue a DELETE request to the URL or API shorthand."""
        response = self._session.delete(self._resolve_url(url))
        self.log_request(response)
        response.raise_for_status()
        # Catch no content responses from some delete actions.
        if response.status_code == 204:
            return []
        return self._response_value(response)

    def upload_creative(self, account_id, file_path):
        """Upload a media creative to the account with ID
        account_id from the local file_path.

        """
        # Thanks to nosklo for his answer on SO:
        # http://stackoverflow.com/a/681182
        boundary = '-----------------------------' + str(int(random.random() * 1e10))
        parts = []

        # Set account ID part.
        parts.append('--' + boundary)
        parts.append('Content-Disposition: form-data; name="account_id"')
        parts.append('')
        parts.append(str(account_id))

        # Set creative contents part.
        parts.append('--' + boundary)
        parts.append('Content-Disposition: form-data; name="userfile"; filename="%s"' % file_path)
        parts.append('Content-Type: %s' % mimetypes.guess_type(file_path)[0] or 'application/octet-stream')
        parts.append('')
        # TODO: catch errors with opening file.
        with open(file_path, 'r') as f:
            parts.append(f.read())

        parts.append('--' + boundary + '--')
        parts.append('')

        body = '\r\n'.join(parts)

        # TODO: Catch errors in attempt to upload.
        headers = {'content-type': 'multipart/form-data; boundary=' + boundary}
        if self.api_path == API_PATH_V1:
            url = self._resolve_url('/a/creative/uploadcreative')
        elif self.api_path == API_PATH_V2:
            url = self._resolve_url('/creative/uploadcreative')
        else:
            raise UnknownAPIFormatError(
                'Unrecognized API path: %s' % self.api_path)
        response = self._session.get(url, headers=headers, data=body, timeout=self.timeout)
        self.log_request(response)
        response.raise_for_status()
        return self._response_value(response)


def client_from_file(file_path='.ox3rc', env=None):
    """Return an instance of ox3apiclient.Client with data from file_path.

    Keyword arguments:
    file_path -- the file to load. Default is '.ox3rc' form current dir.
    env -- the env section to load. Default will be first env section.

    """
    cp = ConfigParser.RawConfigParser()
    cp.read(file_path)

    # Load default env if no env is specified. The default env is just the first
    # env listed.
    if not env:
        env = [e for e in cp.get('ox3apiclient', 'envs').split('\n') if e][0]

    # Required parameters for a ox3apiclient.Client instance.
    required_params = [
        'domain',
        'consumer_key',
        'consumer_secret']

    client_params = {}

    # Load required parameters.
    try:
        for param in required_params:
            client_params[param] = cp.get(env, param)
    except ConfigParser.NoOptionError:
        err_msg = "Missing required option: '%s'" % param
        raise Exception(err_msg)

    client = Client(
        domain=client_params['domain'],
        realm=None,
        consumer_key=client_params['consumer_key'],
        consumer_secret=client_params['consumer_secret'])

    # Load optional parameters.
    optional_params = [
        'callback_url',
        'scheme',
        'request_token_url',
        'access_token_url',
        'authorization_url',
        'api_path',
        'email',
        'password',
        'timeout']

    for param in optional_params:
        try:
            prop = param

            # Prefix private properties with '_'.
            if prop in ['email', 'password']:
                prop = '_%s' % prop

            client.__dict__[prop] = cp.get(env, param)

        except ConfigParser.NoOptionError:
            pass

    return client


# The exposed API has moved to using Client instead of OX3APIClient, but create
# a temporary alias for backwards compatibility.
OX3APIClient = Client
